### PR TITLE
Connection: verify REST API errors

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
Currently only XMLRPC errors for requests coming from WPCOM get added to the "Verified Errors" list.
This PR also adds there the errors that occur if an REST API request to WPCOM fails.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
p9dueE-5zL-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Connect Jetpack.
2. Use the "Jetpack Debug -> Broken Token" plugin to set invalid blog and user tokens.
3. Go to Jetpack Dashboard to trigger the error.
4. Go to "Jetpack Debug - Connection Errors" and confirm an error with the code `unknown_token` shows up in both "Current Unverified Errors" and "Current Verified Errors" sections. The error should contain the `error_type: rest` value.
5. Click on the "Clear all errors" button, reload the page and confirm the errors have disappeared.
6. Run [Jetpack Debug Tool](https://jptools.wordpress.com/debug/) over the site, go back to the "Jetpack Debug -> Connection Errors".
7. Confirm that the error `no_valid_blog_token` was created with `error_type: xmlrpc`. If no errors were created, they might've been blocked by the "gate". Please remove the transients from the database and run the "Jetpack Debug Tool" again:
```
delete from wp_options where option_name like "%connection_error_reporting_gate%";
```

### CODE REVIEW

**Comment by sergeymitr** [comment]
Hi @leogermani.
I know it's been a while since you worked on this, but can you please review this PR when you get a minute to make sure I didn't miss something important.

**Comment by leogermani** [comment]
Hi Sergey,

I gave it a go and overall it seems to work, but it would be nice if you could have a look at some points:

The test you describe in the test instructions works, but this used to generate a XMLRPC error before, and now I only see the REST error. Maybe the gate is preventing one of them from being logged, or there's something else there.

Metadata

I think it's ok to start handling other types of error there, but it would be a good idea to add some more data to the errors so we can make a distinction between them.

For example, be able to differentiate "outgoing REST" errors from the "incoming XMLRPC" error the class current handles.

And then we could/should add conditions to the `handle_verified_errors` method, so we know precisely what we are going to do with each kind of error.

Also, it might helps us not mess them up when limiting the number of error we store (a max of 5 per user_id per error_code). And not mess the gate up...

Docs

In the same spirit, it would be great to update some internal docs (like the big comment at the top of the class) and other places to make it explicit that this is not just about XMLRPC error any more. Also the debug tool.

File: projects/packages/connection/src/class-error-handler.php [hunk 1]
```diff
@@ -199,15 +199,16 @@ public static function get_instance() {
...
 	 *
-	 * @param \WP_Error $error the error object.
-	 * @param boolean   $force Force the report, even if should_report_error is false.
+	 * @param \WP_Error $error  The error object.
+	 * @param boolean   $force  Force the report, even if should_report_error is false.
+	 * @param boolean   $verify_locally Set to 'true' to verify the error locally and skip the WP.com verification.
 	 * @return void
 	 */
-	public function report_error( \WP_Error $error, $force = false ) {
+	public function report_error( \WP_Error $error, $force = false, $verify_locally = false ) {
```
Comment (leogermani) [inline projects/packages/connection/src/class-error-handler.php]: nitpick suggestion. I think `$skip_wpcom_verification` would be a better name for this

Comment (sergeymitr) [reply projects/packages/connection/src/class-error-handler.php]: Good point, renamed: https://github.com/Automattic/jetpack/pull/25694/commits/1153aad07a019971c315ae969b058bdbc055e99d

**Comment by sergeymitr** [comment]
Hi @leogermani, thanks for taking a look!

> For example, be able to differentiate "outgoing REST" errors from the "incoming XMLRPC" error the class current handles.

I added the error type (`xmlrpc` or `rest`) to be stored in the errors: https://github.com/Automattic/jetpack/pull/25694/commits/e4f5d903ee220bd4f936db8131e059697625d499

> And then we could/should add conditions to the `handle_verified_errors` method, so we know precisely what we are going to do with each kind of error.

I don't think we should handle the REST errors differently. They are still legit, and  will still show up the same as XMLRPC ones with those codes.
But being able to do that if the need arises is a good idea 🙂 

> Also, it might helps us not mess them up when limiting the number of error we store (a max of 5 per user_id per error_code). And not mess the gate up...

I didn't modify those behaviors, so they still only rely on error codes and user ID's. The error type (`xmlrpc` or `rest`) isn't taken into account. I believe it doesn't really matter where the error is coming from, as long as we are sure it's valid.

> In the same spirit, it would be great to update some internal docs (like the big comment at the top of the class) and other places to make it explicit that this is not just about XMLRPC error any more. Also the debug tool.

I updated the files, but left intact:
- `/jetpack/v4/verify_xmlrpc_error` because it's still only used for XML-RPC errors
- `jetpack_connection_xmlrpc_errors` and `jetpack_connection_xmlrpc_verified_errors` options - that would require a migration, I think we should take care of it separately not to bloat this PR.

Please take another look and let me know if you notice any issues.

**Review by leogermani — APPROVED** [review]
LGTM!

Thanks for all the changes.

I was wondering if it's worth to store the difference that XMLRPC requests are incoming requests and REST api are outgoing requests... (in case in the future we want to also log outgoing XMLRCP and incoming REST)

**Comment by sergeymitr** [comment]
Thanks @leogermani!

> I was wondering if it's worth to store the difference that XMLRPC requests are incoming requests and REST api are outgoing requests... (in case in the future we want to also log outgoing XMLRCP and incoming REST)

That sounds like a good idea, but we can add it later when we get there 🙂

**Pull request merged**